sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala

```diff
@@ -317,8 +320,18 @@ class JacksonParser(
 
     // SPARK-25040: Disallow empty strings for data types except for string and binary types.
     case VALUE_STRING if parser.getTextLength < 1 =>
-      throw new RuntimeException(
-        s"Failed to parse an empty string for data type ${dataType.catalogString}")
+        // Legacy mode for prior to Spark 3.0.0.
+      if (allowEmptyString) {
+        dataType match {
```

Comment:
If you would match by types that allow empty strings, this could be shorter:
```scala
dataType match {
  case IntergerType | ... if allowEmptyString => null
  case _ => throw new RuntimeException(s"Failed to parse an empty string for data type ${dataType.catalogString}")
}
```

Reply:
Then we need to match types including `ArrayType`, `MapType` etc. The whitelist seems too long.